File: problem9.js
// ## Problem 1: (https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/)

// Given an array of integers nums sorted in ascending order, find the starting and ending position of a given target value.

// Your algorithm's runtime complexity must be in the order of O(log n).

// If the target is not found in the array, return [-1, -1].

// Example 1:

// Input: nums = [5,7,7,8,8,10], target = 8
// Output: [3,4]
// Example 2:

// Input: nums = [5,7,7,8,8,10], target = 6
// Output: [-1,-1]

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
// time complexity:
// space complexity: 
var binarySearchFirst = function(nums, low, high, target){
    while(low <= high){
        let mid = Math.floor(low +(high - low)/2);
        if(nums[mid] === target){
          // if it is 1st index of target?
            if(mid === 0 || nums[mid] !== nums[mid-1]){
                return mid;
            }
            high = mid -1;
        }else if(nums[mid] > target){
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }  
      return -1;
  }
  
// time complexity:
// space complexity: 
var binarySearchLast = function(nums, low, high, target){
    while(low <= high){
    let mid = Math.floor(low +(high - low)/2);
    if(nums[mid] === target){
        // if it is last index of target?
        if(mid === nums.length-1 || nums[mid] !== nums[mid+1]){
            return mid;
        }
        low = mid + 1;
    }else if(nums[mid] > target){
        high = mid - 1;
    } else {
        low = mid + 1;
    }
}  
    return -1;
}

// time complexity:
// space complexity: 
var searchRange = function(nums, target) {
    let n = nums.length
    if(nums === null || n === 0) return [-1,-1];
    if(nums[0] > target || nums[n] < target) return [-1,-1];
    let first = binarySearchFirst(nums, 0, n-1, target);
    if(first == -1) return [-1,-1];
    let last = binarySearchLast(nums, first, n-1, target);
    return [first, last];
    
};

File: problem 9.py
# Time Complexity : O(log n)
# Space Complexity :o(1)
# Did this code successfully run on Leetcode : YES
# Any problem you faced while coding this : NO

# Your code here along with comments explaining your approach in three sentences only

# edge cases like if the array is empty or if the target is not in the range of array return [-1,-1]
# find first occurrence of the target in the array by doing BS, if mid is found and the left of mid is not target then return mid
# else go left and if nums[mid] > target also go left else go right
# similarly for the last occurrence of the target do bs, if mid is found and the right of mid is not target then return mid
# else go right and if nums[mid] > target also go left else go right

class Solution:
    #bs to find the first occurrence of target
    def bsfirst(self,nums: List[int], low: int, high: int, target: int) ->int:
        while(low <= high):
            mid = low + (high - low)//2
            if nums[mid] == target: # if target is found and left of it is not equal to that return that mid
                if mid == 0 or nums[mid] != nums[mid-1]:
                    return mid
                else: # else go left
                    high = mid -1
            elif nums[mid] > target: # if nums[mid] > target go left
                high = mid - 1
            else:   # else go right
                low = mid + 1
        return -1

    #bs to find the last occurrence of target
    def bslast(self,nums: List[int], low: int, high: int, target: int) ->int:
        n = len(nums)
        while(low <= high):
            mid = low + (high - low)//2
            if nums[mid] == target: # if target is found and right of it is not equal to that return that mid
                if mid == n-1 or nums[mid] != nums[mid+1]:
                    return mid
                else: # else go right
                    low = mid +1
            elif nums[mid] > target: # if nums[mid] > target go left
                high = mid - 1
            else: # else go right
                low = mid + 1
        return -1


    def searchRange(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)
        if (n == 0 or nums == None): #if array is null return -1,-1
            return [-1,-1]
        if target < nums[0] or target > nums[n-1]: # if target is not in the range in array return -1,-1
            return [-1,-1]
        
        first = self.bsfirst(nums,0,n-1,target) #bs to find the first occurrence of target
        if(first == -1):
            return [-1,-1]
        last = self.bslast(nums,first,n-1,target) #bs to find the last occurrence of target
        return [first,last]